Sources/Commands/SwiftTestTool.swift

```diff
@@ -196,21 +196,36 @@ public class SwiftTestTool: SwiftTool<TestToolOptions> {
     ///
     /// - Returns: The path to the test binary.
     private func buildTestsIfNeeded(_ options: TestToolOptions) throws -> AbsolutePath {
-        let buildPlan = try BuildPlan(buildParameters: self.buildParameters(), graph: loadPackageGraph())
-        if options.shouldBuildTests {
-            try build(plan: buildPlan, subset: .allIncludingTests)
+        // Check if we need to create the build plan.
+        if try shouldRegenerateManifest() || toolCache.testBinary == nil {
+            let buildPlan = try BuildPlan(buildParameters: buildParameters(), graph: loadPackageGraph())
+            if options.shouldBuildTests {
+                try self.build(plan: buildPlan, subset: .allIncludingTests)
+            }
+            guard let testProduct = buildPlan.buildProducts.first(where: { $0.product.type == .test }) else {
+                throw TestError.testsExecutableNotFound
+            }
+
+            // Go up the folder hierarchy until we find the .xctest bundle.
+            let testBinary = sequence(first: testProduct.binary, next: {
+                $0.isRoot ? nil : $0.parentDirectory
+            }).first(where: {
+                $0.basename.hasSuffix(".xctest")
+            })!
```

Comment:
It uses `!` because that should never happen. It is like using fatalError for an impossible condition. There is probably a better way to express that but that is orthogonal to this PR.

Reply:
That sounds reasonable as long as we test that the xctest extension is guaranteed to be in the path when `testProduct` exists